Test/Cluster: Don't open replica immediately after crash

sentientwaffle opened the pull request:
Reopening the replica immediately after crash was a shortcut that kept all of the cluster's `Replica`s valid. (Then the "crashed" replicas would just not be `tick`'d until they restarted).

But it makes more sense to `open` when the replica is actually restarted.

This allows `Cluster.restart_replica()` to report `open`'s error, rather than `Cluster.crash_replica()`.

## Pre-merge checklist

Performance:

* [x] I am very sure this PR could not affect performance.

sentientwaffle committed:
Test/Cluster: Don't open replica immediately after crash

Reopening the replica immediately after crash was a shortcut that kept all of the cluster's `Replica`s valid.
(Then the "crashed" replicas would just not be `tick`'d until they restarted).

But it makes more sense to `open` when the replica is actually restarted.
- This allows `Cluster.restart_replica()` to report `open`'s error, rather than `Cluster.crash_replica()`.

matklad commented on src/testing/cluster.zig:
```diff
@@ -300,28 +305,28 @@ pub fn ClusterType(comptime StateMachineType: fn (comptime Storage: type, compti
 
+            // Pass the old replica's Time through to the new replica. It will continue to tick while
+            // the replica is crashed, to ensure the clocks don't desyncronize too far to recover.
+            try cluster.open_replica(replica_index, cluster.replicas[replica_index].time);
```
My mental borrow-checker sounds alarm here: I think sometimes Zig sneakily converts by-value parameters into by-immutable-reference, and this might be problematic for the `Time` parameter here, as we are overwriting the whole replica in `open_replica`. 

Is there some sure-fire way to force a copy here?

sentientwaffle replied:
Considering how many times I have accidentally copied a huge type to the stack by assigning it to a variable, this should do the trick:

https://github.com/tigerbeetledb/tigerbeetle/pull/543/commits/9a91d0acdca80c41dbd3039bef8bce91b4c193eb

sentientwaffle committed:
Test/Cluster: Force copy to stack

Since `open_replica` will overwrite the replica, and zig sometimes converts by-value arguments to references.

matklad commented on src/testing/cluster.zig:
```diff
@@ -300,28 +305,29 @@ pub fn ClusterType(comptime StateMachineType: fn (comptime Storage: type, compti
-            const replica = &cluster.replicas[replica_index];
-            const replica_time = replica.time;
-            replica.deinit(cluster.allocator);
+            cluster.replicas[replica_index].deinit(cluster.allocator);
```
Let's add `self.* = undefined` to `Replica.deinit`. I _think_ that would add runtime assertions in debug if we actually try to access that?

matklad replied:
Ah, that won't work, because we do access Replica to tick it's timer :thinking:

sentientwaffle replied:
We could change the `time` that is passed into a `*Time` (similar to how `Storage` and `MessageBus` are not owned by the replica itself).

sentientwaffle replied:
Then `time` would be owned by the `Cluster` (like storage and messagebus).

matklad replied:
Accessing a field of a de-inited struct feels a bit questionable, but probably is OK for simulator. The alternative would be to fold `replica_healf` into `replicas`:

```zig
union(enum) {
  up: Replica,
  down: Time,
}
```

but that feels like too much churn. But maybe that's ok?

matklad replied:
>We could change the time that is passed into a *Time

Yeah, that probably is better. Still in doubt whether it is worth it

matklad reviewed (approved):
I think this LGTM as is, even with our hacky handling of time, but I lean slightly towards doing something more principled there (*Time seems like the best option)

sentientwaffle commented:
Yeah I'll change it over to `*Time`, I've been meaning to change that for a while.
